src/wp-includes/l10n.php

```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
 		$locale = determine_locale();
 	}
 
-	$mo = new MO();
-	if ( ! $mo->import_from_file( $mofile ) ) {
-		$wp_textdomain_registry->set( $domain, $locale, false );
+	// Ensures the correct locale is set as the current one, in case it was filtered.
+	WP_I18n_Translation_Controller::instance()->set_locale( $locale );
 
-		return false;
+	/**
+	 * Filters the preferred file format for translation files.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $preferred_format Preferred file format. Possible values: 'php', 'mo'. Default: 'php'.
+	 */
+	$preferred_format = apply_filters( 'translation_file_format', 'php' );
+	if ( ! in_array( $preferred_format, array( 'php', 'mo' ), true ) ) {
+		$preferred_format = 'php';
 	}
```

Comment:
What's the rationale for introducing this filter? What would be a realistic use-case for it? If there is one, I'm not opposed, but I would advise to rather be cautious on introducing new filters when we don't have to as it limits what changes we can make later without breaking BC.

Reply:
Related to your other question:

> Similar to the other new filter, what's the rationale / use-case for introducing the `load_translation_file` filter?

I was thinking a long time about these two filters and what the best approach is.

`load_textdomain_mofile` is historically the one filter to change the file path, but plugins expect only MO files here.

Hence the introduction of the more generic `load_translation_file` filter which could also receive a PHP file.

I still opted to call the `load_textdomain_mofile` filter for PHP files as well, but that's probably not the safest approach. So I'll just remove it again.

Reply:
@swissspidy Why would people want to do that? In what case would MO files have a benefit over PHP? I just want to make sure we allow it for good reasons, not just because someone wants that (without a good reason).

The other related question: How would that work with shipping translations? Wouldn't this preference need to equally apply when eventually getting language packs from .org (i.e. a parameter needs to be sent to .org which format is preferred)?

Reply:
Dotorg sends both formats in a ZIP file, it does not support conditionally sending only half.

Originally the filter was added to choose between MO/JSON/PHP. Now someone might want to prefer PHP if they have some concerns about that format or if they are not even using dotorg and only have MO files anyway, skipping a lot of file_exists checks

Reply:
I would prefer that, we can always introduce filters when needed, but we can't remove them. I prefer that over adding filters for a _potential_ need. It helps avoid unnecessary filters that we then still have to maintain forever.

The `file_exists()` checks is a good point, though I'd still prefer to remove for now and park this for a later conversation.

Reply:
It would be relevant for https://github.com/WordPress/wordpress-develop/pull/5306#issuecomment-1805349307, so let's await that discussion...